src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,920 @@
+                                            statement)))).WithAdditionalAnnotations(Simplifier.Annotation);
+                        }
+                    case SyntaxKind.SelectClause:
+                        // This is not the latest Select in the Query Expression
+                        // There is a QueryBody with the Continuation as a parent.
+                        var selectClause = (SelectClauseSyntax)node;
+                        var identifier = ((QueryBodySyntax)selectClause.Parent).Continuation.Identifier;
+                        return AddToBlockTop(CreateLocalDeclarationStatement(identifier, selectClause.Expression), statement);
+                    default:
+                        throw new ArgumentException($"Unexpected node kind {node.Kind().ToString()}");
+                }
+            }
+
+            private bool TryConvertInternal(QueryExpressionProcessingInfo queryExpressionProcessingInfo, out DocumentUpdateInfo documentUpdateInfo)
+            {
+                // (from a in b select a); 
+                SyntaxNode parent = _source.WalkUpParentheses().Parent;
+
+                switch (parent.Kind())
+                {
+                    // return from a in b select a;
+                    case SyntaxKind.ReturnStatement:
+                        return TryConvertIfInReturnStatement((ReturnStatementSyntax)parent, queryExpressionProcessingInfo, out documentUpdateInfo);
+                    // foreach(var x in from a in b select a)
+                    case SyntaxKind.ForEachStatement:
+                        return TryConvertIfInForEach((ForEachStatementSyntax)parent, queryExpressionProcessingInfo, out documentUpdateInfo);
+                    // (from a in b select a).ToList(), (from a in b select a).Count(), etc.
+                    case SyntaxKind.SimpleMemberAccessExpression:
+                        return TryConvertIfInMemberAccessExpression((MemberAccessExpressionSyntax)parent, out documentUpdateInfo);
+                }
+
+                documentUpdateInfo = default;
+                return false;
+            }
+
+            /// <summary>
+            /// Checks if the location of the query expression allows to convert it at least to a local function.
+            /// It still does not guarantees that the conversion can be performed. There can be bail outs of later stages.
+            /// </summary>
+            /// <returns></returns>
+            private bool CanTryConvertToLocalFunction()
+            {
+                SyntaxNode currentNode = _source;
+                while (currentNode != null)
+                {
+                    switch (currentNode.Kind())
+                    {
+                        case SyntaxKind.CatchClause:
+                        case SyntaxKind.WhenClause:
+                        case SyntaxKind.ArrowExpressionClause: // This one can be considered for support later.
```

Comment:
>case SyntaxKind.ArrowExpressionClause [](start = 24, length = 37)

We are trying to identify unsafe parents here, but it is easy miss one. Perhaps we should use an opposite approach and look if parents are "safe". For example, we can go up the parents chain and check that we can reach StatementSyntax before reaching MemberDeclarationSyntax and along the way we only go through MemberDeclarationSyntax and ArgumentSyntax. I think it is better to be conservative here and enable more scenarios later. For example, it looks like this function will allow refactoring a query which is inside a query clause. That is probably fine to do if the query is a source of a top level from, or a source of a join, but feels questionable for a query inside where clause

Reply:
There was a typo in the comment meant to say ExpressionSyntax instead of MemberDeclarationSyntax. Fixed 

---
In reply to: [178377999](https://github.com/dotnet/roslyn/pull/25362#discussion_r178377999) [](ancestors = 178377999)

Reply:
actually, we should allow essentially more:
currentNode is ExpressionSyntax ||
                        currentNode is ArgumentSyntax ||
                        currentNode is ArgumentListSyntax ||
                        currentNode is EqualsValueClauseSyntax ||
                        currentNode is VariableDeclaratorSyntax ||
                        currentNode is VariableDeclarationSyntax ||
                        currentNode is MemberAccessExpressionSyntax

---
In reply to: [178378332](https://github.com/dotnet/roslyn/pull/25362#discussion_r178378332) [](ancestors = 178378332,178377999)